File: atcoder/abc/2/1/abc211/d.rb
MOD = 10 ** 9 + 7
def gs; gets.chomp; end
def gi; gets.to_i; end
def gli; gets.split.map(&:to_i); end
def cyesno(x); x ? :Yes : :No; end
def uyesno(x); x ? :YES : :NO; end
def array_to_hash_by_count(c); Hash.new(0).tap { |x| c.each { |z| x[z] += 1 } }; end

def bfs(adj, src, dist, paths, n)
  visited = Array.new(n, false)
  dist[src] = 0
  paths[src] = 1
  q = Array.new(1, src)

  visited[src] = true

  while q.any? do
    curr = q[0]
    q.shift

    # For all neighbors of current vertex do:
    adj[curr].each do |x|
      # if the current vertex is not yet
      # visited, then push it to the queue.
      unless visited[x]
        q.push(x)
        visited[x] = true
      end
      
      # check if there is a better path. 
      if dist[x] > dist[curr] + 1
        dist[x] = dist[curr] + 1
        paths[x] = paths[curr]
      # additional shortest paths found
      elsif dist[x] == dist[curr] + 1
        paths[x] += paths[curr]
      end
    end
  end
end

def find_shortest_paths(adj, s, n)
  dist = Array.new(n, Float::INFINITY)
  paths = Array.new(n, 0)
  bfs(adj, s, dist, paths, n)
  return paths
end
 
def add_edge(adj, u, v)
  adj[u].push(v)
  adj[v].push(u)
end

def main
  n, m = gli
  adj = []
  for i in (0..n - 1)
    adj[i] = []
  end
  m.times do
    u, v = gli
    add_edge(adj, u - 1, v - 1)
  end
  puts find_shortest_paths(adj, 0, n)[-1] % MOD
end

main

File: atcoder/abc/2/1/abc211/d.py
from collections import deque
from sys import maxsize as INT_MAX
 
def BFS(adj: list, src: int, dist: list, paths: list, n: int):
    visited = [False] * n
    dist[src] = 0
    paths[src] = 1
 
    q = deque()
    q.append(src)
    visited[src] = True
    while q:
        curr = q[0]
        q.popleft()
 
        for x in adj[curr]:
            if not visited[x]:
                q.append(x)
                visited[x] = True
 
            if dist[x] > dist[curr] + 1:
                dist[x] = dist[curr] + 1
                paths[x] = paths[curr]
 
            elif dist[x] == dist[curr] + 1:
                paths[x] += paths[curr]
 
def findShortestPaths(adj: list, s: int, n: int):
    dist = [INT_MAX] * n
    paths = [0] * n
    BFS(adj, s, dist, paths, n)
    print(paths[-1] % 1000000007)
 
def addEdge(adj: list, u: int, v: int):
    adj[u].append(v)
    adj[v].append(u)
 
if __name__ == "__main__":
    n, m = map(int, input().split())
    adj = [0] * n
    for i in range(n):
        adj[i] = []
    for i in range(m):
        x, y = map(int, input().split())
        addEdge(adj, x - 1, y - 1)
    findShortestPaths(adj, 0, n)
 
